### PR TITLE
[WIP] implementing startup checks

### PR DESCRIPTION
**Pull request by melkio**
@Particular/azure-maintainers I need your feedback about changes to ITopology.Initialize signature to have container reference into AzureServiceBusTransport class. Any other way to obtain the same behaviour?
After this first check, I go ahead with refactoring.

Take a look at my comment at the beginning of startu check callback.

NOTE: ApiComparer test fails

### CODE REVIEW

**Comment by SeanFeldman** [comment]
@melkio now that you've been introduced to the secret order of ApiComparer, that test can go green :)

File: src/Transport/Topology/ITopology.cs [hunk 1]
```diff
@@ -6,7 +6,7 @@ namespace NServiceBus.AzureServiceBus
 
     public interface ITopology {
 
-        void Initialize(SettingsHolder settings);
+        IResolveTransportParts Initialize(SettingsHolder settings);
```
Comment (SeanFeldman) [inline src/Transport/Topology/ITopology.cs]: Is Command–Query separation principle a concern here?

File: src/Transport/Config/AzureServiceBusTransport.cs [hunk 2]
```diff
@@ -20,7 +21,25 @@ protected override TransportReceivingConfigurationResult ConfigureForReceiving(T
...
+                {
+                    // Useless for me: when we resolve INamespacePartitioningStrategy to apply startup check, we have already applied checks defined into ctor of concrete classes. Right?
+                    var namespacePartitioningStrategy = Container.Resolve<INamespacePartitioningStrategy>();
+                    var namespaces = context.Settings.Get<List<string>>(WellKnownConfigurationKeys.Topology.Addressing.Partitioning.Namespaces);
+                    var endpointName = context.Settings.Get<EndpointName>();
+
+                    var configuredNamespaces = namespacePartitioningStrategy.GetNamespaces(endpointName.ToString(), PartitioningIntent.Creating);
+
+                    if (namespaces.Count != configuredNamespaces.Count())
+                        return Task.FromResult(StartupCheckResult.Failed("..."));
```
Comment (SeanFeldman) [inline src/Transport/Config/AzureServiceBusTransport.cs]: "..." needs to be more descriptive :)

File: src/Transport/Config/AzureServiceBusTransport.cs [hunk 3]
```diff
@@ -20,7 +21,25 @@ protected override TransportReceivingConfigurationResult ConfigureForReceiving(T
...
+
+                    var namespaceManager  = Container.Resolve<IManageNamespaceManagerLifeCycle>()
+                        .Get(context.ConnectionString);
+                    if (context.Settings.Get<bool>(WellKnownConfigurationKeys.Core.CreateTopology) && !namespaceManager.HasManageRights)
+                        return Task.FromResult(StartupCheckResult.Failed("..."));
```
Comment (SeanFeldman) [inline src/Transport/Config/AzureServiceBusTransport.cs]: "..." needs to be more descriptive

File: src/Transport/Config/AzureServiceBusTransport.cs [hunk 4]
```diff
@@ -20,7 +21,25 @@ protected override TransportReceivingConfigurationResult ConfigureForReceiving(T
...
+                    var namespaces = context.Settings.Get<List<string>>(WellKnownConfigurationKeys.Topology.Addressing.Partitioning.Namespaces);
+                    var endpointName = context.Settings.Get<EndpointName>();
+
+                    var configuredNamespaces = namespacePartitioningStrategy.GetNamespaces(endpointName.ToString(), PartitioningIntent.Creating);
+
+                    if (namespaces.Count != configuredNamespaces.Count())
+                        return Task.FromResult(StartupCheckResult.Failed("..."));
+
+                    var namespaceManager  = Container.Resolve<IManageNamespaceManagerLifeCycle>()
+                        .Get(context.ConnectionString);
```
Comment (SeanFeldman) [inline src/Transport/Config/AzureServiceBusTransport.cs]: Might be wrong here, but wouldn't we need to run this on all the registered `namespaces`?

File: src/Transport/Connectivity/NamespaceManagerAdapter.cs [hunk 5]
```diff
@@ -19,6 +19,22 @@ public NamespaceManagerAdapter(NamespaceManager manager)
...
+        public bool HasManageRights
+        {
+            get
+            {
+                try
+                {
+                    _manager.GetQueues();
+                    return true;
+                }
+                catch
```
Comment (SeanFeldman) [inline src/Transport/Connectivity/NamespaceManagerAdapter.cs]: We need to be careful here in order not to confuse `UnauthorizedAccessException` for anything else.

**Comment by melkio** [comment]
@SeanFeldman this first commit was only to receive feedback about "design". I totally agree with your comments. I'll fix all the things and I'll do a lot of refactoring later if you agree with ITopology changes

I'm working on a possible alternative...a new PR is coming :)

**Comment by melkio** [comment]
I don't like this implementation.

If you agree, I'd close this PR. What do you think @Particular/azure-maintainers?

**Comment by SeanFeldman** [comment]
Go for it @melkio 
(close PR + delete the branch)

**Pull request closed without merging**